pkg/services/publicdashboards/service/service.go

```diff
@@ -253,8 +256,48 @@ func (pd *PublicDashboardServiceImpl) buildMetricRequest(ctx context.Context, da
 	// determine safe resolution to query data at
 	safeInterval, safeResolution := pd.getSafeIntervalAndMaxDataPoints(reqDTO, ts)
 	for i := range queries {
+		datasourceType := queries[i].Get("datasource").Get("type").MustString()
+		pd.log.Info(fmt.Sprintf("datasource type %s", datasourceType))
+
 		queries[i].Set("intervalMs", safeInterval)
 		queries[i].Set("maxDataPoints", safeResolution)
+
+		// replace template variables
+		// this is a POC, not intended to get merged
+		// this loop is replacing all occurrences of ${var} and $var inside rawSql and expr
+		// using the default values declared in the original dashboard
+		// we should have different services for each datasource type (like postgres, prometheus, graphite)
+		// and different methods for each query type (textbox, custom, list, multivalue)
+		for _, varObj := range dashboard.Data.Get("templating").Get("list").MustArray() {
+			variable := simplejson.NewFromAny(varObj)
+			var value string
+
+			if variable.Get("multi").MustBool() {
+				valuesObj := variable.Get("current").Get("value").MustArray()
+				var valuesArray []string
+
+				for _, valueObj := range valuesObj {
+					valuesArray = append(valuesArray, valueObj.(string))
+				}
+				value = strings.Join(valuesArray[:], ",")
+			} else {
+				value = variable.Get("current").Get("value").MustString()
+			}
+
+			name := variable.Get("name").MustString()
+
+			// some DS uses rawSQL
+			rawSql := queries[i].Get("rawSql").MustString()
+			rawSql = strings.Replace(rawSql, fmt.Sprintf("${%s}", name), value, -1)
+			rawSql = strings.Replace(rawSql, fmt.Sprintf("$%s", name), value, -1)
```

Comment:
We have some more complex syntaxes that support formatting option as well:
https://grafana.com/docs/grafana/latest/dashboards/variables/variable-syntax/ 

```${var_name:<format>}```

here are the formats: https://github.com/grafana/grafana/blob/main/public/app/features/templating/formatRegistry.ts 

We need a separate TemplateInterpolationService I think which each data source can use to interpolate the query. But I guess that is not going to work well for external backend data sources. Maybe each data source can in plugin.json define how to interpolate
strings by default? tricky stuff.